### PR TITLE
feat: new data generator

### PR DESCRIPTION
**Pull request by pauldix**
I didn't quite get to where I wanted to today, but here's a draft for an early look. This builds and passes lints. But I need to look through everything again and clean up cruft, update the README, add a spec that shows all the possible functionality, and then write up what the follow on work will be.

However, I'm quite happy with where this landed in terms of performance. Here's a preview of what I generated with the tracing spec example and the storage cardinality example. Each agent that is simulated generating data is single threaded. So the way to take advantage of multiple cores is with many agents. Using just one here's the rough number from my laptop:

```
Tracing Spec generates 451,500/sec per core (agent)
Storage Cardinality Example: 2,250,075/sec per core
```

And then I updated the storage cardinality example to see what kind of scale I could get. Set with cardinality of **300k** per agent and **100 agents** for a total cardinality of **30M** uses **203.7MB** of RAM. Generated data and wrote to no-op at a rate of **13,125,000** per second.

Here are some snippets of what the generated line protocol look like:

```
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=1,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=3781610i 1635531216713744000
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=2,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=4215103i 1635531216713744000
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=3,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=7930129i 1635531216713744000
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=4,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=386008i 1635531216713744000
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=5,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=946454i 1635531216713744000
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=6,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=4615919i 1635531216713744000
storage_usage_bucket_cardinality,bucket_id=1_6pfYxLJAGRrVm7c,env=whatever-environment-1,host=storage-0,hostname=0,node_id=0,org_id=1_c3t6wLzsAsdfdJ8,partition_id=7,role=storage,url=http://127.0.0.1:6060/metrics/usage gauge=5748844i 1635531216713744000
```

```
traces,host=server-1,service=service-1,span_id=d10d6190-3e72-43f1-aa5b-98913bf80c63,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=177.68025338056714 1635531267075005000
traces,host=server-2,service=service-2,span_id=824151a0-5b66-4728-ab62-407c6d464017,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=191.87250112959055 1635531267075005000
traces,host=server-3,service=service-3,span_id=824151a0-5b66-4728-ab62-407c6d464017,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=419.61109662616957 1635531267075005000
traces,host=server-4,service=service-4,span_id=7d3ebde9-33e5-4730-b7a3-7207791db1ae,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=432.1260987984159 1635531267075005000
traces,host=server-5,service=service-5,span_id=7d3ebde9-33e5-4730-b7a3-7207791db1ae,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=389.03458528786683 1635531267075005000
traces,host=server-6,service=service-6,span_id=925c9826-9c15-4cb4-bcd1-7c7eaa2198e4,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=108.5402030631698 1635531267075005000
traces,host=server-7,service=service-7,span_id=925c9826-9c15-4cb4-bcd1-7c7eaa2198e4,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=413.1363278415766 1635531267075005000
traces,host=server-8,service=service-8,span_id=9cefe63d-5fbd-4e23-941c-ad03beb1bdaa,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=161.22643508691857 1635531267075005000
traces,host=server-9,service=service-9,span_id=9cefe63d-5fbd-4e23-941c-ad03beb1bdaa,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=357.7788132345887 1635531267075005000
traces,host=server-10,service=service-10,span_id=4eb4a433-566f-4c74-a7fc-ebf3006da8ff,trace_id=43e251b8-a0f8-4da2-b240-8ae68c2c8a96 timing=158.86915057869254 1635531267075005000
```

### CODE REVIEW

File: iox_data_generator/src/specification.rs [hunk 1]
```diff
@@ -186,77 +147,48 @@ pub struct MeasurementSpec {
...
-    ///   each tag will have a different GUID.
-    pub value: String,
-    /// The number of tags with this configuration that should be created.
-    /// Default value is 1. If specified, use `{{tag_id}}` in this tag's
-    /// `name` to create unique tags.
+pub struct TagPairSpec {
+    /// The tag key. If `count` is specified, the id of the tag will be automatically
+    /// appended to the end of the key to ensure it is unique.
+    pub key: String,
+    /// The template the generate the tag value
```
Comment (carols10cents) [inline iox_data_generator/src/specification.rs:179]: Typo:

```suggestion
    /// The template to generate the tag value
```

**Review by carols10cents — CHANGES_REQUESTED** [review]
Yup, this is a lot simpler without the random seed functionality 😂 The `LineToGenerator` is nice too!

I'm getting a panic when I try to run the benchmarks because the spec used for the benchmark has some values that aren't used anymore. Other than that, the rest are nits you can take or leave

File: iox_data_generator/src/measurement.rs [hunk 2]
```diff
@@ -1,331 +1,379 @@
...
+        let fields = spec
             .fields
             .iter()
             .map(|field_spec| {
-                FieldGeneratorSet::new::<T>(
-                    &agent_name,
-                    agent_id,
-                    measurement_id,
+                // TODO: add field count here
+                //                let count = field_spec.count.unwrap_or(1);
```
[the remark this answers is not in the document]
Comment (pauldix) [reply iox_data_generator/src/measurement.rs:129]: Ah, meant to take care of this in this PR. Will fix!

File: iox_data_generator/src/agent.rs [hunk 3]
```diff
@@ -71,65 +73,67 @@ pub struct Agent<T: DataGenRng> {
...
-                    spec,
-                    &seed,
-                    &agent_tags,
-                    execution_start_time,
-                )
+        generated_tag_sets: &GeneratedTagSets,
+    ) -> Result<Vec<Self>> {
+        let agent_count = agent_spec.count.unwrap_or(1);
+
+        let agents: Vec<_> = (0..agent_count)
```
[the remark this answers is not in the document]
Comment (pauldix) [reply iox_data_generator/src/agent.rs:90]: Ah yes, this was unintentional. Will fix so they all start from 1 and will update Fields to work.

File: iox_data_generator/schemas/full_example.toml [hunk 4]
```diff
@@ -1,79 +1,172 @@
...
 ]
 
 [[agents]]
-# create this many agents
-count = 1
+# Create this many agents. Agents are single threaded so the way to get more paralellism is to
+# increase the number of agents.
+count = 2
+# each agent will generate a sampling of data every 10ms. If running from a historical point in
+# time, the agents will push d
```
Comment (carols10cents) [inline iox_data_generator/schemas/full_example.toml:111]: Looks like an unfinished comment here?

File: iox_data_generator/schemas/full_example.toml [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+# new_agent_measurement-1,agent-name=another_example f1=f 1635968398968189000
+# new_agent_measurement-2,agent-name=another_example f1=f 1635968398968189000
+
+name = "full_example"
+
+# Values are automatically generated before the agents are intialized. They generate tag key/value pairs
+# with the name of the value as the tag key and the evaluated template as the value. These pairs
+# can be shared across tagsets and used in the agents as pre-generated data.
+[[values]]
+# The name is appears as the tag key in generated tag pairs and is used later when specifying tag_sets.
```
Comment (carols10cents) [inline iox_data_generator/schemas/full_example.toml:40]: Nit:

```suggestion
# The name appears as the tag key in generated tag pairs and is used later when specifying tag_sets.
```

**Review by carols10cents — APPROVED** [review]
Looking great! A few tiny comment suggestions, but overall running the benches works now and everything looks like it was addressed!

**Pull request merged**